File: motor/notas.py
"""Extracción y formateo de notas al pie (§5.5). Las notas se extraen ANTES
de dividir capítulos (D6); el backlink del capítulo se asigna tras el split."""

import re

from motor.limpieza import limpiar_texto_html
from motor.modelo import Chapter, Nota

RE_SECCION_NOTAS = re.compile(
    r'<(section|div)\b[^>]*(?:id|class)="footnotes?"[^>]*>(.*?)</\1>',
    re.DOTALL | re.IGNORECASE,
)
RE_ITEM_NOTA = re.compile(
    r'<li\b[^>]*id="fn(\d+)"[^>]*>(.*?)</li>',
    re.DOTALL | re.IGNORECASE,
)
RE_VINCULO_NOTA_PANDOC = re.compile(
    r'<a\b[^>]*href="#fn(\d+)"[^>]*>.*?</a>',
    re.DOTALL,
)
RE_VINCULO_REGRESO = re.compile(
    r'<a\b[^>]*href="#fnref[^"]*"[^>]*>.*?</a>',
    re.DOTALL,
)
RE_NOTA_LEGACY = re.compile(r'\(NT(\d+)\)')
RE_ANCLA_REGRESO = re.compile(r'↩︎?')

ARCHIVO_NOTAS = 'notas_Finales.xhtml'


def formatear_llamada(num: int) -> str:
    """Llamada inline: <a href="notas_Finales.xhtml#nt01" id="rf01">…"""
    num_fmt = f'{num:02d}'
    return (
        f'<a href="{ARCHIVO_NOTAS}#nt{num_fmt}" id="rf{num_fmt}">'
        f'<sup>❮{num_fmt}❯</sup></a>'
    )


def _reemplazar_notas_legacy(html: str) -> str:
    """Reemplaza referencias legacy (NT##) por llamadas en formato ePub."""

    def reemplazar(match: re.Match) -> str:
        return formatear_llamada(int(match.group(1)))

    return RE_NOTA_LEGACY.sub(reemplazar, html)


def extraer_notas(html: str) -> tuple[str, list[Nota]]:
    """Extrae la sección de notas (pandoc) del HTML y la devuelve formateada
    como lista de Nota; las llamadas inline se actualizan al formato ePub."""
    notas: list[Nota] = []

    match = RE_SECCION_NOTAS.search(html)
    if match:
        contenido_raw = match.group(2)
        html = html[:match.start()] + html[match.end():]

        for num_str, html_nota in RE_ITEM_NOTA.findall(contenido_raw):
            num = int(num_str)
            html_nota = RE_VINCULO_REGRESO.sub('', html_nota)
            html_nota = RE_ANCLA_REGRESO.sub('', html_nota)
            html_nota = re.sub(r'^<p>', '', html_nota.strip())
            html_nota = re.sub(r'</p>$', '', html_nota.strip())
            html_nota = limpiar_texto_html(html_nota.strip())
            notas.append(Nota(num=num, texto=html_nota))

        def reemplazar_llamada(match: re.Match) -> str:
            return formatear_llamada(int(match.group(1)))

        html = RE_VINCULO_NOTA_PANDOC.sub(reemplazar_llamada, html)

    html = _reemplazar_notas_legacy(html)
    return html, notas


def asignar_capitulos(notas: list[Nota], capitulos: list[Chapter], start_num: int = 1) -> None:
    """Asigna a cada nota el capítulo que contiene su llamada (id="rfNN"),
    guardando su nombre de archivo real (§5.8); por defecto el primer
    capítulo si no está referenciada."""
    for nota in notas:
        referencia = f'id="rf{nota.num:02d}"'
        for num, capitulo in enumerate(capitulos, start=start_num):
            if referencia in capitulo.html_cuerpo:
                nota.cap_num = num
                nota.cap_archivo = (
                    capitulo.archivo or f'C{num:02d}.xhtml'
                )
                break
        else:
            nota.cap_num = start_num
            nota.cap_archivo = f'C{start_num:02d}.xhtml'


def formatear_nota(nota: Nota) -> str:
    """Formatea una nota como div ePub con su backlink al capítulo."""
    num_fmt = f'{nota.num:02d}'
    archivo = nota.cap_archivo
    if archivo is None:
        archivo = f'C{nota.cap_num if nota.cap_num is not None else 1:02d}.xhtml'
    return (
        f'<div class="nota">\n'
        f' <p id="nt{num_fmt}">\n'
        f'   <a href="{archivo}#rf{num_fmt}"><sup>❮{num_fmt}❯</sup> {nota.texto}</a>\n'
        f' </p>\n'
        f'</div>'
    )

File: app_flutter/lib/motor/notas.dart
import 'modelo.dart';
import 'limpieza.dart';

final _reSeccionNotas = RegExp(
  r'<(section|div)\b[^>]*(?:id|class)="footnotes?"[^>]*>(.*?)</\1>',
  dotAll: true, caseSensitive: false,
);
final _reItemNota = RegExp(
  r'<li\b[^>]*id="fn(\d+)"[^>]*>(.*?)</li>',
  dotAll: true, caseSensitive: false,
);
final _reVinculoNotaPandoc = RegExp(
  r'<a\b[^>]*href="#fn(\d+)"[^>]*>.*?</a>',
  dotAll: true,
);
final _reVinculoRegreso = RegExp(
  r'<a\b[^>]*href="#fnref[^"]*"[^>]*>.*?</a>',
  dotAll: true,
);
final _reNotaLegacy = RegExp(r'\(NT(\d+)\)');
final _reNotaMdDef = RegExp(r'<p>\s*\[\^(\d+)\]:\s*(.*?)</p>', dotAll: true);
final _reNotaMdLlamada = RegExp(r'\[\^(\d+)\]');
final _reAnclaRegreso = RegExp(r'↩︎?');
final _reImg = RegExp(r'<img\b[^>]*?/?>', dotAll: true, caseSensitive: false);

const archivoNotas = 'notas_Finales.xhtml';

String formatearLlamada(int num) {
  String numFmt = num.toString().padLeft(2, '0');
  return '<a href="$archivoNotas#nt$numFmt" id="rf$numFmt">'
         '<sup>❮$numFmt❯</sup></a>';
}

String _reemplazarNotasLegacy(String html) {
  return html.replaceAllMapped(_reNotaLegacy, (m) => formatearLlamada(int.parse(m[1]!)));
}

String _procesarNotasMarkdown(String html, List<Nota> notas) {
  for (final m in _reNotaMdDef.allMatches(html)) {
    notas.add(Nota(
      num: int.parse(m[1]!),
      texto: limpiarTextoHtml(m[2]!.trim()),
    ));
  }
  html = html.replaceAll(_reNotaMdDef, '');
  return html.replaceAllMapped(_reNotaMdLlamada, (m) => formatearLlamada(int.parse(m[1]!)));
}

class ExtraerNotasResult {
  final String html;
  final List<Nota> notas;
  ExtraerNotasResult(this.html, this.notas);
}

ExtraerNotasResult extraerNotas(String html) {
  List<Nota> notas = [];
  final match = _reSeccionNotas.firstMatch(html);
  
  if (match != null) {
    String contenidoRaw = match.group(2)!;
    html = html.substring(0, match.start) + html.substring(match.end);

    for (final m in _reItemNota.allMatches(contenidoRaw)) {
      int num = int.parse(m[1]!);
      String htmlNota = m[2]!;
      htmlNota = htmlNota.replaceAll(_reVinculoRegreso, '');
      htmlNota = htmlNota.replaceAll(_reAnclaRegreso, '');
      htmlNota = htmlNota.trim().replaceAll(RegExp(r'^<p>'), '').replaceAll(RegExp(r'</p>$'), '');
      htmlNota = limpiarTextoHtml(htmlNota.trim());
      notas.add(Nota(num: num, texto: htmlNota));
    }

    html = html.replaceAllMapped(_reVinculoNotaPandoc, (m) => formatearLlamada(int.parse(m[1]!)));
  }

  html = _procesarNotasMarkdown(html, notas);
  html = _reemplazarNotasLegacy(html);
  return ExtraerNotasResult(html, notas);
}

void asignarCapitulos(List<Nota> notas, List<Chapter> capitulos, {int startNum = 1}) {
  for (var nota in notas) {
    String referencia = 'id="rf${nota.num.toString().padLeft(2, '0')}"';
    bool encontrada = false;
    
    for (int i = 0; i < capitulos.length; i++) {
      int num = startNum + i;
      if (capitulos[i].htmlCuerpo.contains(referencia)) {
        nota.capNum = num;
        nota.capArchivo = capitulos[i].archivo ?? 'C${num.toString().padLeft(2, '0')}.xhtml';
        encontrada = true;
        break;
      }
    }
    
    if (!encontrada) {
      nota.capNum = startNum;
      nota.capArchivo = 'C${startNum.toString().padLeft(2, '0')}.xhtml';
    }
  }
}

String _textoConImagenes(String texto, int num) {
  final numeros = _reImg.allMatches(texto).toList();
  if (numeros.isEmpty) return texto;

  List<String> partes = [];
  int posicion = 0;
  
  for (int i = 0; i < numeros.length; i++) {
    int indice = i + 1;
    final match = numeros[i];
    String sufijo = indice > 1 ? '-$indice' : '';
    String fragmento = texto.substring(posicion, match.start).trim();
    
    if (fragmento.isNotEmpty && partes.isNotEmpty) partes.add('<br/><br/>');
    if (fragmento.isNotEmpty) partes.add(fragmento);
    
    partes.add('<img src="../Images/nota-${num.toString().padLeft(2, '0')}$sufijo.jpg" alt=""/>');
    posicion = match.end;
  }

  String resto = texto.substring(posicion).trim();
  if (resto.isNotEmpty) {
    partes.add('<br/><br/>');
    partes.add(resto);
  }
  return partes.join(' ');
}

String formatearNota(Nota nota) {
  String numFmt = nota.num.toString().padLeft(2, '0');
  String archivo = nota.capArchivo ?? 'C${(nota.capNum ?? 1).toString().padLeft(2, '0')}.xhtml';
  String texto = _textoConImagenes(nota.texto, nota.num);
  
  String div = '<div class="nota">\n'
               ' <p id="nt$numFmt">\n'
               '   <a href="$archivo#rf$numFmt"><sup>❮$numFmt❯</sup> $texto</a>\n'
               ' </p>\n'
               '</div>';
               
  if (texto.contains('<img')) {
    return '<hr class="sigil_split_marker" />\n$div\n<hr class="sigil_split_marker" />';
  }
  return div;
}
